File: core/systems/embedded-systems/lab/embedded-core-dsa/cpp-embedded-edge/circular_buffer.cpp
#include <array>
#include <stdexcept>
#include <cassert>

template <typename T, size_t N>
class CircularBuffer {
    std::array<T, N> buffer;
    size_t head = 0;
    size_t tail = 0;
    bool full = false;

public:
    void push(const T& item) {
        buffer[head] = item;
        head = (head + 1) % N;
        if (full) tail = (tail + 1) % N;
        full = (head == tail);
    }

    T pop() {
        if (empty()) throw std::runtime_error("Buffer empty");
        T val = buffer[tail];
        full = false;
        tail = (tail + 1) % N;
        return val;
    }

    bool empty() const { return !full && (head == tail); }
    bool is_full() const { return full; }
};

// Test Case
void test_circular_buffer() {
    CircularBuffer<int, 3> cb;
    cb.push(10); cb.push(20); cb.push(30);
    assert(cb.is_full());
    assert(cb.pop() == 10);
    cb.push(40);
    assert(cb.pop() == 20);
}

File: core/systems/embedded-systems/lab/embedded-core-dsa/python-embedded-edge/circular_buffer.py
from collections import deque

class CircularBuffer:
    def __init__(self, size):
        self.buffer = deque(maxlen=size)

    def push(self, data):
        self.buffer.append(data)

    def pop(self):
        if self.buffer:
            return self.buffer.popleft()
        else:
            raise IndexError("Buffer is empty")

    def is_full(self):
        return len(self.buffer) == self.buffer.maxlen

    def is_empty(self):
        return len(self.buffer) == 0

# Test
def test_circular_buffer():
    cb = CircularBuffer(3)
    cb.push(1)
    cb.push(2)
    cb.push(3)
    assert cb.is_full()
    assert cb.pop() == 1
    cb.push(4)
    assert cb.pop() == 2
    assert not cb.is_empty()

test_circular_buffer()
